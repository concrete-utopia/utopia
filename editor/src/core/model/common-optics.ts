import type { EditorStorePatched } from '../../components/editor/store/editor-state'
import {
  modifyUnderlyingForOpenFile,
  withUnderlyingTargetFromEditorState,
} from '../../components/editor/store/editor-state'
import type { JSXElementChild } from '../shared/element-template'
import type { Optic } from '../shared/optics/optics'
import { traversal } from '../shared/optics/optics'
import type { ElementPath } from '../shared/project-file-types'

export function forElementOptic(target: ElementPath): Optic<EditorStorePatched, JSXElementChild> {
  function from(store: EditorStorePatched): Array<JSXElementChild> {
    return withUnderlyingTargetFromEditorState(target, store.editor, [], (_, element) => {
      return [element]
    })
  }
  function update(
    store: EditorStorePatched,
    modify: (child: JSXElementChild) => JSXElementChild,
  ): EditorStorePatched {
    return {
      ...store,
      editor: modifyUnderlyingForOpenFile(target, store.editor, modify),
    }
  }
  return traversal(from, update)
}
