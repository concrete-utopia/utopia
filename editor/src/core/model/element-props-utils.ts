import { type AllElementProps } from '../../components/editor/store/editor-state'
import { type ElementPath, type PropertyPath } from '../shared/project-file-types'
import * as PP from '../shared/property-path'
import * as EP from '../shared/element-path'
import * as ObjectPath from 'object-path'

export function getElementPropValue(
  allElementProps: AllElementProps,
  target: ElementPath,
  propPath: PropertyPath,
): unknown {
  const propsForElement = allElementProps[EP.toString(target)] ?? {}
  return ObjectPath.get(propsForElement, PP.getElements(propPath))
}

export function elementHasValueForProp(
  allElementProps: AllElementProps,
  target: ElementPath,
  propPath: PropertyPath,
): boolean {
  return getElementPropValue(allElementProps, target, propPath) != null
}
