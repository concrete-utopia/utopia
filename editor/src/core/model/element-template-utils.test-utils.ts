import { MOCK_NEXT_GENERATED_UIDS, MOCK_NEXT_GENERATED_UIDS_IDX } from '../shared/uid-utils'

export function FOR_TESTS_setNextGeneratedUid(nextUid: string): void {
  MOCK_NEXT_GENERATED_UIDS.current = [nextUid]
  MOCK_NEXT_GENERATED_UIDS_IDX.current = 0
}

export function FOR_TESTS_setNextGeneratedUids(uids: Array<string>): void {
  MOCK_NEXT_GENERATED_UIDS.current = uids
  MOCK_NEXT_GENERATED_UIDS_IDX.current = 0
}

export function FOR_TESTS_CLEAR_MOCK_NEXT_GENERATED_UIDS(): void {
  MOCK_NEXT_GENERATED_UIDS.current = []
  MOCK_NEXT_GENERATED_UIDS_IDX.current = 0
}

// automatic cleanup after tests
afterEach(() => {
  FOR_TESTS_CLEAR_MOCK_NEXT_GENERATED_UIDS()
})
