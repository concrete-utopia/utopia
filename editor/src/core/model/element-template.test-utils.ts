import { isUtopiaJSXComponent, UtopiaJSXComponent } from '../shared/element-template'
import { isParseSuccess } from '../shared/project-file-types'
import { parseCode } from '../workers/parser-printer/parser-printer'

export function getComponentFromCode(componentName: string, code: string): UtopiaJSXComponent {
  const parseResult = parseCode('test.jsx', code, null, new Set())
  if (isParseSuccess(parseResult)) {
    for (const topLevelElement of parseResult.topLevelElements) {
      if (isUtopiaJSXComponent(topLevelElement) && topLevelElement.name === componentName) {
        return topLevelElement
      }
    }
    throw new Error(`Could not find component ${componentName}`)
  } else {
    throw new Error(`Not a parse success: ${parseResult.type}`)
  }
}
