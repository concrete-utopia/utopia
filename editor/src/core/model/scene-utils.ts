import Hash from 'object-hash'
import {
  SceneMetadata,
  StaticElementPath,
  PropertyPath,
  isTextFile,
  isParseSuccess,
} from '../shared/project-file-types'
import {
  UtopiaJSXComponent,
  utopiaJSXComponent,
  jsxElement,
  JSXElement,
  jsExpressionValue,
  isJSXElement,
  JSXElementChild,
  TopLevelElement,
  isUtopiaJSXComponent,
  JSXAttributes,
  defaultPropsParam,
  jsExpressionOtherJavaScript,
  ElementInstanceMetadataMap,
  jsxAttributesFromMap,
  ElementInstanceMetadata,
  walkElements,
  emptyComments,
} from '../shared/element-template'
import * as EP from '../shared/element-path'
import * as PP from '../shared/property-path'
import {
  eitherToMaybe,
  applicative5Either,
  flatMapEither,
  isLeft,
  applicative6Either,
  right,
} from '../shared/either'
import { memoize } from '../shared/memoize'
import fastDeepEqual from 'fast-deep-equal'
import {
  getModifiableJSXAttributeAtPath,
  jsxSimpleAttributeToValue,
} from '../shared/jsx-attributes'
import { stripNulls } from '../shared/array-utils'
import { UTOPIA_UID_KEY } from './utopia-constants'
import { getContentsTreeFileFromString, ProjectContentTreeRoot } from '../../components/assets'
import { getUtopiaJSXComponentsFromSuccess } from './project-file-utils'
import { generateConsistentUID, generateUID, getUtopiaID } from '../shared/uid-utils'
import { emptySet } from '../shared/set-utils'

export const PathForSceneComponent = PP.create('component')
export const PathForSceneDataUid = PP.create('data-uid')
export const PathForSceneDataLabel = PP.create('data-label')
export const PathForSceneFrame = PP.create('style')

export const BakedInStoryboardUID = 'utopia-storyboard-uid'
export const BakedInStoryboardVariableName = 'storyboard'

export const EmptyUtopiaCanvasComponent = convertScenesToUtopiaCanvasComponent([])

export const PathForSceneProps = PP.create('props')
export const PathForSceneStyle = PP.create('style')

export function createSceneUidFromIndex(sceneIndex: number): string {
  return `scene-${sceneIndex}`
}

export function mapScene(scene: SceneMetadata): JSXElement {
  const sceneProps = jsxAttributesFromMap({
    component: jsExpressionOtherJavaScript(
      scene.component ?? 'null',
      `return ${scene.component}`,
      [],
      null,
      {},
    ),
    props: jsExpressionValue(scene.props, emptyComments),
    style: jsExpressionValue(scene.frame, emptyComments),
    'data-uid': jsExpressionValue(scene.uid, emptyComments),
    'data-label': jsExpressionValue(scene.label, emptyComments),
  })
  return jsxElement('Scene', scene.uid, sceneProps, [])
}

export function unmapScene(element: JSXElementChild): SceneMetadata | null {
  if (!isJSXElement(element) || element.name.baseVariable !== 'Scene') {
    return null
  }
  return eitherToMaybe(
    applicative6Either(
      (component, props, style, label, uid) => {
        const scene: SceneMetadata = {
          uid: uid,
          component: component,
          props: props,
          frame: style,
          ...(label == null ? {} : { label: label }),
        }
        return scene
      },
      getSimpleAttributeAtPathCustom(element.props, PP.create('component')),
      getSimpleAttributeAtPathCustom(element.props, PP.create('props')),
      getSimpleAttributeAtPathCustom(element.props, PP.create('style')),
      getSimpleAttributeAtPathCustom(element.props, PP.create('layout')),
      getSimpleAttributeAtPathCustom(element.props, PP.create('data-label')),
      getSimpleAttributeAtPathCustom(element.props, PP.create('data-uid')),
    ),
  )
}

export function convertScenesToUtopiaCanvasComponent(
  scenes: Array<SceneMetadata>,
): UtopiaJSXComponent
export function convertScenesToUtopiaCanvasComponent(scenes: null): null
export function convertScenesToUtopiaCanvasComponent(
  scenes: Array<SceneMetadata> | null,
): UtopiaJSXComponent | null
export function convertScenesToUtopiaCanvasComponent(
  scenes: Array<SceneMetadata> | null,
): UtopiaJSXComponent | null {
  if (scenes == null) {
    return null
  }
  return utopiaJSXComponent(
    BakedInStoryboardVariableName,
    false,
    'var',
    'block',
    null,
    [],
    jsxElement(
      'Storyboard',
      BakedInStoryboardUID,
      jsxAttributesFromMap({ 'data-uid': jsExpressionValue(BakedInStoryboardUID, emptyComments) }),
      scenes.map(mapScene),
    ),
    null,
    false,
    emptyComments,
  )
}

export function createSceneFromComponent(
  filePath: string,
  componentImportedAs: string,
  uid: string,
): JSXElement {
  const sceneProps = jsxAttributesFromMap({
    [UTOPIA_UID_KEY]: jsExpressionValue(uid, emptyComments),
    style: jsExpressionValue(
      {
        position: 'absolute',
        left: 0,
        top: 0,
        width: 375,
        height: 812,
      },
      emptyComments,
    ),
  })
  const hash = Hash({
    fileName: filePath,
    name: componentImportedAs,
    props: jsxAttributesFromMap({}),
  })
  const componentUID = generateConsistentUID(hash)
  return jsxElement('Scene', uid, sceneProps, [
    jsxElement(
      componentImportedAs,
      componentUID,
      jsxAttributesFromMap({
        [UTOPIA_UID_KEY]: jsExpressionValue(componentUID, emptyComments),
      }),
      [],
    ),
  ])
}

export function createStoryboardElement(scenes: Array<JSXElement>, uid: string): JSXElement {
  const storyboardProps = jsxAttributesFromMap({
    [UTOPIA_UID_KEY]: jsExpressionValue(uid, emptyComments),
  })
  return jsxElement('Storyboard', uid, storyboardProps, scenes)
}

export function convertUtopiaCanvasComponentToScenes(
  utopiaCanvasComponent: UtopiaJSXComponent | null,
): Array<SceneMetadata> | null {
  if (utopiaCanvasComponent == null) {
    return null
  }
  const rootElement = utopiaCanvasComponent.rootElement
  if (!isJSXElement(rootElement) || rootElement.name.baseVariable !== 'Storyboard') {
    throw new Error('the root element must be a Storyboard component')
  }
  const firstChildrenScenes: Array<SceneMetadata> = stripNulls(rootElement.children.map(unmapScene))
  return firstChildrenScenes
}

const convertScenesAndTopLevelElementsToUtopiaCanvasComponentMemoized = memoize(
  (scenes: Array<SceneMetadata>, topLevelElements: Array<TopLevelElement>) => [
    ...topLevelElements,
    convertScenesToUtopiaCanvasComponent(scenes),
  ],
  { equals: fastDeepEqual }, // delete me as soon as possible
)

export function convertScenesAndTopLevelElementsToUtopiaCanvasComponent(
  scenes: Array<SceneMetadata>,
  topLevelElements: Array<UtopiaJSXComponent>,
): Array<UtopiaJSXComponent>
export function convertScenesAndTopLevelElementsToUtopiaCanvasComponent(
  scenes: Array<SceneMetadata>,
  topLevelElements: Array<TopLevelElement>,
): Array<TopLevelElement>
export function convertScenesAndTopLevelElementsToUtopiaCanvasComponent(
  scenes: Array<SceneMetadata>,
  topLevelElements: Array<TopLevelElement>,
): Array<TopLevelElement> {
  return convertScenesAndTopLevelElementsToUtopiaCanvasComponentMemoized(scenes, topLevelElements)
}

export function fishOutUtopiaCanvasFromTopLevelElements(
  topLevelElements: Array<TopLevelElement>,
): UtopiaJSXComponent | null {
  return (
    topLevelElements.find((e): e is UtopiaJSXComponent => {
      return isUtopiaJSXComponent(e) && e.name === BakedInStoryboardVariableName
    }) ?? null
  )
}

function getSimpleAttributeAtPathCustom(attributes: JSXAttributes, path: PropertyPath) {
  const getAttrResult = getModifiableJSXAttributeAtPath(attributes, path)
  return flatMapEither((attr) => {
    const simpleValue = jsxSimpleAttributeToValue(attr)
    if (isLeft(simpleValue) && attr.type === 'ATTRIBUTE_OTHER_JAVASCRIPT') {
      return right(attr.javascript)
    } else {
      return simpleValue
    }
  }, getAttrResult)
}

export function sceneMetadata(
  uid: string,
  component: string | null,
  props: { [key: string]: any },
  frame: {
    left: number
    top: number
    width: number
    height: number
  },
  label?: string,
): SceneMetadata {
  let scene: SceneMetadata = {
    uid: uid,
    component: component,
    frame: frame,
    props: props,
  }

  if (label != null) {
    // This is annoying, but if we don't do this then we'll explicitly print
    // `undefined` in the label field, meaning the JSON won't parse
    scene.label = label
  }

  return scene
}

export function getStoryboardUID(openComponents: UtopiaJSXComponent[]): string | null {
  const possiblyStoryboard = openComponents.find(
    (component) => component.name === BakedInStoryboardVariableName,
  )
  if (possiblyStoryboard != null) {
    return getUtopiaID(possiblyStoryboard.rootElement)
  }
  return null
}

export function getStoryboardElementPath(
  projectContents: ProjectContentTreeRoot,
  openFile: string | null | undefined,
): StaticElementPath | null {
  if (openFile != null) {
    const file = getContentsTreeFileFromString(projectContents, openFile)
    if (file != null && isTextFile(file) && isParseSuccess(file.fileContents.parsed)) {
      const possiblyStoryboard = getUtopiaJSXComponentsFromSuccess(file.fileContents.parsed).find(
        (component) => component.name === BakedInStoryboardVariableName,
      )
      if (possiblyStoryboard != null) {
        const uid = getUtopiaID(possiblyStoryboard.rootElement)
        return EP.elementPath([EP.staticElementPath([uid])])
      }
    }
  }
  return null
}
