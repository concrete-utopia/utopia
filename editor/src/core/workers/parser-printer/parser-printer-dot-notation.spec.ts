import { applyPrettier } from './prettier-utils'
import {
  clearParseResultUniqueIDs,
  testParseCode,
  testParseThenPrint,
} from './parser-printer.test-utils'
import { BakedInStoryboardUID } from '../../model/scene-utils'
import { foldParsedTextFile } from '../../shared/project-file-types'

describe('JSX parser', () => {
  it('should parse out dot-notation element names', () => {
    const code = applyPrettier(
      `
import * as React from "react";
import * as Utopia from "utopia-api";
export var App = props => {
  return (
    <Utopia.View data-uid="aaa">
      {<div data-uid="bbb" />}
    </Utopia.View>
  )
}
export var storyboard = (props) => {
  return (
    <Storyboard data-uid='${BakedInStoryboardUID}'>
      <Scene
        style={{ left: 0, top: 0, width: 400, height: 400 }}
        component={App}
        layout={{ layoutSystem: 'pinSystem' }}
        props={{ layout: { bottom: 0, left: 0, right: 0, top: 0 } }}
        data-uid='scene-aaa'
      />
    </Storyboard>
  )
}
`,
      false,
    ).formatted
    const parseResult = clearParseResultUniqueIDs(testParseCode(code))
    foldParsedTextFile(
      (failure) => {
        fail(failure)
      },
      (success) => {
        expect(success.imports).toMatchSnapshot()
        expect(success.topLevelElements).toMatchSnapshot()
      },
      (unparsed) => {
        fail(unparsed)
      },
      parseResult,
    )
  })
  it('dot-notated elements cycle fully back and forth', () => {
    const originalCode = `
import * as React from "react";
import * as Utopia from "utopia-api";
import { Scene, Storyboard, View } from "utopia-api";
export var App = props => {
  return (
    <Utopia.View data-uid="aaa">
      <View data-uid='bbb' />
    </Utopia.View>
  )
}
export var storyboard = (props) => {
  return (
    <Storyboard data-uid='${BakedInStoryboardUID}'>
      <Scene
        style={{ left: 0, top: 0, width: 400, height: 400 }}
        component={App}
        layout={{ layoutSystem: 'pinSystem' }}
        props={{ layout: { bottom: 0, left: 0, right: 0, top: 0 } }}
        data-uid='scene-aaa'
      />
    </Storyboard>
  )
}`
    const printedCode = `import * as React from 'react'
import { Scene, Storyboard, View } from 'utopia-api'
import * as Utopia from 'utopia-api'
export var App = (props) => {
  return (
    <Utopia.View data-uid='aaa'>
      <View data-uid='bbb' />
    </Utopia.View>
  )
}
export var storyboard = (props) => {
  return (
    <Storyboard data-uid='${BakedInStoryboardUID}'>
      <Scene
        style={{ left: 0, top: 0, width: 400, height: 400 }}
        component={App}
        layout={{ layoutSystem: 'pinSystem' }}
        props={{ layout: { bottom: 0, left: 0, right: 0, top: 0 } }}
        data-uid='scene-aaa'
      />
    </Storyboard>
  )
}
`
    testParseThenPrint(originalCode, printedCode)
  })
})
