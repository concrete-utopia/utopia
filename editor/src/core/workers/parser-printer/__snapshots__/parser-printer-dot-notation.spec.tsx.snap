// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSX parser should parse out dot-notation element names 1`] = `
Object {
  "react": Object {
    "importedAs": "React",
    "importedFromWithin": Array [],
    "importedWithName": null,
  },
  "utopia-api": Object {
    "importedAs": "Utopia",
    "importedFromWithin": Array [],
    "importedWithName": null,
  },
}
`;

exports[`JSX parser should parse out dot-notation element names 2`] = `
Array [
  Object {
    "importStarAs": true,
    "importWithName": false,
    "imports": Array [],
    "module": "react",
    "rawCode": "import * as React from 'react'",
    "type": "IMPORT_STATEMENT",
  },
  Object {
    "rawCode": "
",
    "type": "UNPARSED_CODE",
  },
  Object {
    "importStarAs": true,
    "importWithName": false,
    "imports": Array [],
    "module": "utopia-api",
    "rawCode": "import * as Utopia from 'utopia-api'",
    "type": "IMPORT_STATEMENT",
  },
  Object {
    "rawCode": "

",
    "type": "UNPARSED_CODE",
  },
  Object {
    "arbitraryJSBlock": null,
    "blockOrExpression": "block",
    "declarationSyntax": "var",
    "functionWrapping": Array [],
    "isFunction": true,
    "name": "App",
    "params": Array [
      Object {
        "boundParam": Object {
          "defaultExpression": null,
          "paramName": "props",
          "type": "REGULAR_PARAM",
        },
        "dotDotDotToken": false,
        "type": "PARAM",
      },
    ],
    "propsUsed": Array [],
    "returnStatementComments": Object {
      "leadingComments": Array [],
      "trailingComments": Array [],
    },
    "rootElement": Object {
      "children": Array [
        Object {
          "comments": Object {
            "leadingComments": Array [],
            "trailingComments": Array [],
          },
          "definedElsewhere": Array [
            "React",
            "utopiaCanvasJSXLookup",
          ],
          "elementsWithin": Object {
            "bbb": Object {
              "children": Array [],
              "name": Object {
                "baseVariable": "div",
                "propertyPath": Object {
                  "propertyElements": Array [],
                },
              },
              "props": Array [
                Object {
                  "comments": Object {
                    "leadingComments": Array [],
                    "trailingComments": Array [],
                  },
                  "key": "data-uid",
                  "type": "JSX_ATTRIBUTES_ENTRY",
                  "value": Object {
                    "comments": Object {
                      "leadingComments": Array [],
                      "trailingComments": Array [],
                    },
                    "type": "ATTRIBUTE_VALUE",
                    "uid": "",
                    "value": "bbb",
                  },
                },
              ],
              "type": "JSX_ELEMENT",
              "uid": "",
            },
          },
          "javascriptWithUIDs": "<div data-uid='bbb' />;",
          "originalJavascript": "<div data-uid='bbb' />",
          "params": Array [],
          "sourceMap": Object {
            "file": "code.tsx",
            "mappings": "OAMQA;AAAAA;AAAAA",
            "names": Array [
              "<",
            ],
            "sources": Array [
              "code.tsx",
            ],
            "sourcesContent": Array [
              "import * as React from 'react'
import * as Utopia from 'utopia-api'

export var App = (props) => {
  return (
    <Utopia.View data-uid='aaa'>
      {<div data-uid='bbb' />}
    </Utopia.View>
  )
}

export var storyboard = (props) => {
  return (
    <Storyboard data-uid='utopia-storyboard-uid'>
      <Scene
        style={{ left: 0, top: 0, width: 400, height: 400 }}
        data-uid='scene-aaa'
      >
        <App
          data-uid='app-entity'
          style={{ bottom: 0, left: 0, right: 0, top: 0 }}
        />
      </Scene>
    </Storyboard>
  )
}
",
            ],
            "version": 3,
          },
          "transpiledJavascript": "return utopiaCanvasJSXLookup(\\"bbb\\", {
  callerThis: this
});",
          "type": "ATTRIBUTE_OTHER_JAVASCRIPT",
          "uid": "",
        },
      ],
      "name": Object {
        "baseVariable": "Utopia",
        "propertyPath": Object {
          "propertyElements": Array [
            "View",
          ],
        },
      },
      "props": Array [
        Object {
          "comments": Object {
            "leadingComments": Array [],
            "trailingComments": Array [],
          },
          "key": "data-uid",
          "type": "JSX_ATTRIBUTES_ENTRY",
          "value": Object {
            "comments": Object {
              "leadingComments": Array [],
              "trailingComments": Array [],
            },
            "type": "ATTRIBUTE_VALUE",
            "uid": "",
            "value": "aaa",
          },
        },
      ],
      "type": "JSX_ELEMENT",
      "uid": "",
    },
    "type": "UTOPIA_JSX_COMPONENT",
    "usedInReactDOMRender": false,
  },
  Object {
    "rawCode": "

",
    "type": "UNPARSED_CODE",
  },
  Object {
    "definedElsewhere": Array [
      "React",
      "Storyboard",
      "Scene",
      "App",
      "utopiaCanvasJSXLookup",
      "utopiaCanvasBlockRanToEnd",
      "utopiaCanvasEarlyReturnResult",
      "utopiaCanvasEarlyReturnVoid",
    ],
    "definedWithin": Array [
      "storyboard",
    ],
    "elementsWithin": Object {},
    "javascript": "export var storyboard = (props) => {
  return (
    <Storyboard data-uid='utopia-storyboard-uid'>
      <Scene
        style={{ left: 0, top: 0, width: 400, height: 400 }}
        data-uid='scene-aaa'
      >
        <App
          data-uid='app-entity'
          style={{ bottom: 0, left: 0, right: 0, top: 0 }}
        />
      </Scene>
    </Storyboard>
  )
}",
    "params": Array [],
    "sourceMap": Object {
      "file": "code.tsx",
      "mappings": "AAWOA,CAACC,MAAIC;AAAAA,MAAAA,UAAcC,GAAQE,KAADD,IAAIE;AAClCC,WACEC,oBAACC,UAADD;AAAYE,kBAASC;AAArBH,OACEA,oBAACI,KAADJ;AACEK,MAAAA,KAAKT,EAAEU;AAAEC,QAAAA,IAAIC,EAAEC,CAARH;AAAWI,QAAAA,GAAGF,EAAEC,CAAhBH;AAAmBK,QAAAA,KAAKH,EAAEI,GAA1BN;AAA+BO,QAAAA,MAAML,EAAEI;AAAvCN,OADTN;AAEEE,kBAASC;AAFXH,OAIEA,oBAACc,GAADd;AACEE,kBAASC,YADXH;AAEEK,MAAAA,KAAKT,EAAEU;AAAES,QAAAA,MAAMP,EAAEC,CAAVH;AAAaC,QAAAA,IAAIC,EAAEC,CAAnBH;AAAsBU,QAAAA,KAAKR,EAAEC,CAA7BH;AAAgCI,QAAAA,GAAGF,EAAEC;AAArCH;AAFTN,MAJFA,CADFA,CADFD;AAaFkB,GAdWvB;;AAAAA;AAAAA;AAAAA;AAcXuB,CAdMzB",
      "names": Array [
        " ",
        "var",
        "storyboard",
        "props",
        "=",
        ">",
        "
",
        "return",
        "<",
        "Storyboard",
        "data",
        "'",
        "Scene",
        "style",
        "{",
        "left",
        ":",
        "0",
        "top",
        "width",
        "400",
        "height",
        "App",
        "bottom",
        "right",
        "}",
      ],
      "sources": Array [
        "code.tsx",
      ],
      "sourcesContent": Array [
        "import * as React from 'react'
import * as Utopia from 'utopia-api'

export var App = (props) => {
  return (
    <Utopia.View data-uid='aaa'>
      {<div data-uid='bbb' />}
    </Utopia.View>
  )
}

export var storyboard = (props) => {
  return (
    <Storyboard data-uid='utopia-storyboard-uid'>
      <Scene
        style={{ left: 0, top: 0, width: 400, height: 400 }}
        data-uid='scene-aaa'
      >
        <App
          data-uid='app-entity'
          style={{ bottom: 0, left: 0, right: 0, top: 0 }}
        />
      </Scene>
    </Storyboard>
  )
}
",
      ],
      "version": 3,
    },
    "statements": Array [
      Object {
        "assignments": Array [
          Object {
            "leftHandSide": Object {
              "defaultExpression": Object {
                "comments": Object {
                  "leadingComments": Array [],
                  "trailingComments": Array [],
                },
                "definedElsewhere": Array [
                  "React",
                  "Storyboard",
                  "Scene",
                  "App",
                ],
                "elementsWithin": Object {},
                "javascriptWithUIDs": "(props) => {
  return (
    <Storyboard data-uid='utopia-storyboard-uid'>
      <Scene
      style={{ left: 0, top: 0, width: 400, height: 400 }}
      data-uid='scene-aaa'>

        <App
        data-uid='app-entity'
        style={{ bottom: 0, left: 0, right: 0, top: 0 }} />

      </Scene>
    </Storyboard>);

};",
                "originalJavascript": "(props) => {
  return (
    <Storyboard data-uid='utopia-storyboard-uid'>
      <Scene
        style={{ left: 0, top: 0, width: 400, height: 400 }}
        data-uid='scene-aaa'
      >
        <App
          data-uid='app-entity'
          style={{ bottom: 0, left: 0, right: 0, top: 0 }}
        />
      </Scene>
    </Storyboard>
  )
}",
                "params": Array [
                  Object {
                    "boundParam": Object {
                      "defaultExpression": null,
                      "paramName": "props",
                      "type": "REGULAR_PARAM",
                    },
                    "dotDotDotToken": false,
                    "type": "PARAM",
                  },
                ],
                "sourceMap": Object {
                  "file": "code.tsx",
                  "mappings": "OAWyBA,SAAWC;AACjCC,SACEC;AAAYC,gBAAZD;AAAAA,KACEA,oBAACE,KAADF;AACEG,WAAOC;AAAEC,UAAIC,EAANF;AAAWG,SAAGD,EAAdF;AAAmBI,WAAKF,EAAxBF;AAA+BK,YAAMH,EAAtCF;AAACA,KADTJ;AAEEC,gBAFFD;AAAAA,KAIEA,oBAACU,GAADV;AACEC,gBAASU,YADXX;AAEEG,WAAOC;AAAEQ,YAAMN,EAARF;AAAaC,UAAIC,EAAjBF;AAAsBS,WAAKP,EAA3BF;AAAgCG,SAAGD,EAF5CN;AAESI;AAFTJ,IAJFA,CADFA,CADFD;AADqBe,CAACjB",
                  "names": Array [
                    "props",
                    "
",
                    "return",
                    "<",
                    "data",
                    "Scene",
                    "style",
                    "{",
                    "left",
                    ":",
                    "top",
                    "width",
                    "height",
                    "App",
                    "'",
                    "bottom",
                    "right",
                    "(",
                  ],
                  "sources": Array [
                    "code.tsx",
                  ],
                  "sourcesContent": Array [
                    "import * as React from 'react'
import * as Utopia from 'utopia-api'

export var App = (props) => {
  return (
    <Utopia.View data-uid='aaa'>
      {<div data-uid='bbb' />}
    </Utopia.View>
  )
}

export var storyboard = (props) => {
  return (
    <Storyboard data-uid='utopia-storyboard-uid'>
      <Scene
        style={{ left: 0, top: 0, width: 400, height: 400 }}
        data-uid='scene-aaa'
      >
        <App
          data-uid='app-entity'
          style={{ bottom: 0, left: 0, right: 0, top: 0 }}
        />
      </Scene>
    </Storyboard>
  )
}
",
                  ],
                  "version": 3,
                },
                "transpiledJavascript": "return props => {
  return React.createElement(Storyboard, {
    \\"data-uid\\": \\"utopia-storyboard-uid\\"
  }, React.createElement(Scene, {
    style: {
      left: 0,
      top: 0,
      width: 400,
      height: 400
    },
    \\"data-uid\\": \\"scene-aaa\\"
  }, React.createElement(App, {
    \\"data-uid\\": \\"app-entity\\",
    style: {
      bottom: 0,
      left: 0,
      right: 0,
      top: 0
    }
  })));
};",
                "type": "ATTRIBUTE_OTHER_JAVASCRIPT",
                "uid": "",
              },
              "paramName": "storyboard",
              "type": "REGULAR_PARAM",
            },
            "rightHandSide": Object {
              "comments": Object {
                "leadingComments": Array [],
                "trailingComments": Array [],
              },
              "definedElsewhere": Array [
                "React",
                "Storyboard",
                "Scene",
                "App",
              ],
              "elementsWithin": Object {},
              "javascriptWithUIDs": "(props) => {
  return (
    <Storyboard data-uid='utopia-storyboard-uid'>
      <Scene
      style={{ left: 0, top: 0, width: 400, height: 400 }}
      data-uid='scene-aaa'>

        <App
        data-uid='app-entity'
        style={{ bottom: 0, left: 0, right: 0, top: 0 }} />

      </Scene>
    </Storyboard>);

};",
              "originalJavascript": "(props) => {
  return (
    <Storyboard data-uid='utopia-storyboard-uid'>
      <Scene
        style={{ left: 0, top: 0, width: 400, height: 400 }}
        data-uid='scene-aaa'
      >
        <App
          data-uid='app-entity'
          style={{ bottom: 0, left: 0, right: 0, top: 0 }}
        />
      </Scene>
    </Storyboard>
  )
}",
              "params": Array [
                Object {
                  "boundParam": Object {
                    "defaultExpression": null,
                    "paramName": "props",
                    "type": "REGULAR_PARAM",
                  },
                  "dotDotDotToken": false,
                  "type": "PARAM",
                },
              ],
              "sourceMap": Object {
                "file": "code.tsx",
                "mappings": "OAWyBA,SAAWC;AACjCC,SACEC;AAAYC,gBAAZD;AAAAA,KACEA,oBAACE,KAADF;AACEG,WAAOC;AAAEC,UAAIC,EAANF;AAAWG,SAAGD,EAAdF;AAAmBI,WAAKF,EAAxBF;AAA+BK,YAAMH,EAAtCF;AAACA,KADTJ;AAEEC,gBAFFD;AAAAA,KAIEA,oBAACU,GAADV;AACEC,gBAASU,YADXX;AAEEG,WAAOC;AAAEQ,YAAMN,EAARF;AAAaC,UAAIC,EAAjBF;AAAsBS,WAAKP,EAA3BF;AAAgCG,SAAGD,EAF5CN;AAESI;AAFTJ,IAJFA,CADFA,CADFD;AADqBe,CAACjB",
                "names": Array [
                  "props",
                  "
",
                  "return",
                  "<",
                  "data",
                  "Scene",
                  "style",
                  "{",
                  "left",
                  ":",
                  "top",
                  "width",
                  "height",
                  "App",
                  "'",
                  "bottom",
                  "right",
                  "(",
                ],
                "sources": Array [
                  "code.tsx",
                ],
                "sourcesContent": Array [
                  "import * as React from 'react'
import * as Utopia from 'utopia-api'

export var App = (props) => {
  return (
    <Utopia.View data-uid='aaa'>
      {<div data-uid='bbb' />}
    </Utopia.View>
  )
}

export var storyboard = (props) => {
  return (
    <Storyboard data-uid='utopia-storyboard-uid'>
      <Scene
        style={{ left: 0, top: 0, width: 400, height: 400 }}
        data-uid='scene-aaa'
      >
        <App
          data-uid='app-entity'
          style={{ bottom: 0, left: 0, right: 0, top: 0 }}
        />
      </Scene>
    </Storyboard>
  )
}
",
                ],
                "version": 3,
              },
              "transpiledJavascript": "return props => {
  return React.createElement(Storyboard, {
    \\"data-uid\\": \\"utopia-storyboard-uid\\"
  }, React.createElement(Scene, {
    style: {
      left: 0,
      top: 0,
      width: 400,
      height: 400
    },
    \\"data-uid\\": \\"scene-aaa\\"
  }, React.createElement(App, {
    \\"data-uid\\": \\"app-entity\\",
    style: {
      bottom: 0,
      left: 0,
      right: 0,
      top: 0
    }
  })));
};",
              "type": "ATTRIBUTE_OTHER_JAVASCRIPT",
              "uid": "",
            },
            "type": "JS_ASSIGNMENT",
          },
        ],
        "declarationKeyword": "var",
        "type": "JS_ASSIGNMENT_STATEMENT",
        "uid": "",
      },
    ],
    "transpiledJavascript": "return (() => {
  var storyboard = props => {
    return React.createElement(Storyboard, {
      \\"data-uid\\": \\"utopia-storyboard-uid\\"
    }, React.createElement(Scene, {
      style: {
        left: 0,
        top: 0,
        width: 400,
        height: 400
      },
      \\"data-uid\\": \\"scene-aaa\\"
    }, React.createElement(App, {
      \\"data-uid\\": \\"app-entity\\",
      style: {
        bottom: 0,
        left: 0,
        right: 0,
        top: 0
      }
    })));
  };

  return utopiaCanvasBlockRanToEnd({
    storyboard: storyboard
  });
})();",
    "type": "ARBITRARY_JS_BLOCK",
    "uid": "",
  },
  Object {
    "rawCode": "
",
    "type": "UNPARSED_CODE",
  },
]
`;
